From 1fafc52980efae820acf1aac82c772612bcdc1c3 Mon Sep 17 00:00:00 2001
From: xtex <xtexchooser@duck.com>
Date: Sun, 25 May 2025 14:36:48 +0800
Subject: [PATCH 1/2] Fix schema generator for Python 3.14
X-Developer-Signature: v=1; a=openpgp-sha256; l=2157; i=xtexchooser@duck.com;
 h=from:subject; bh=RzYNJ/oJkvJrXLzO3uryWxFFXExmp9XrAca/Ea0a8Js=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDBlGR5w8j5zbvHHj41XlDxPmWrnobHprLSh7oXt9aL799
 cSEh2U6HaUsDGJcDLJiiixFhg3erDrp/KLLymVh5rAygQxh4OIUgIlUH2dkeHTRP6w26FCWVGfT
 wysXPRsz+u5lZ6Zfym/fzXd28c2oTwz/K8u2LlO/eetMt67z0YvhPn/PzC/TWR368md/1r45s8u
 juAE=
X-Developer-Key: i=xtexchooser@duck.com; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Signed-off-by: xtex <xtexchooser@duck.com>
---
 generate_self_schema.py | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/generate_self_schema.py b/generate_self_schema.py
index 623acba58dd5..5c73741319a6 100644
--- a/generate_self_schema.py
+++ b/generate_self_schema.py
@@ -30,6 +30,17 @@ except ImportError:
     TypesUnionType = TypingUnionType
     UnionType = TypingUnionType
 
+try:
+    import annotationlib
+
+    def get_annotations(obj):
+        return {k: annotationlib.ForwardRef(v) for k, v in annotationlib.get_annotations(obj, format=annotationlib.Format.STRING).items()}
+
+
+except ImportError:
+    def get_annotations(obj):
+        return obj.__annotations__
+
 
 THIS_DIR = Path(__file__).parent
 SAVE_PATH = THIS_DIR / 'src' / 'self_schema.py'
@@ -121,7 +132,7 @@ def type_dict_schema(  # noqa: C901
     required_keys = getattr(typed_dict, '__required_keys__', set())
     fields = {}
 
-    for field_name, field_type in typed_dict.__annotations__.items():
+    for field_name, field_type in get_annotations(typed_dict).items():
         required = field_name in required_keys
         schema = None
         fr_arg = None
@@ -194,7 +205,9 @@ def all_literal_values(type_: type[core_schema.Literal]) -> list[any]:
 
 
 def eval_forward_ref(type_: Any) -> Any:
-    if sys.version_info < (3, 12, 4):
+    if sys.version_info >= (3, 14, 0):
+        return type_.evaluate(globals=core_schema.__dict__)
+    elif sys.version_info < (3, 12, 4):
         return type_._evaluate(core_schema.__dict__, None, recursive_guard=set())
     else:
         return type_._evaluate(core_schema.__dict__, None, type_params=set(), recursive_guard=set())
@@ -208,7 +221,7 @@ def main() -> None:
 
     choices = {}
     for s in schema_union.__args__:
-        type_ = s.__annotations__['type']
+        type_ = get_annotations(s)['type']
         m = re.search(r"Literal\['(.+?)']", type_.__forward_arg__)
         assert m, f'Unknown schema type: {type_}'
         key = m.group(1)

base-commit: 34147034f64bff6bd7a240e1475c4faacc3c9b4e
-- 
2.49.0

